File: LeetCode/3.LengthOfLongestSubstring.js
/**
 * LeetCode 3. Longest Substring Without Repeating Characters
 * 
 * Runtime: 96 ms
 * Memory Usage: 40.5 MB
 * 
 * [ "dvdf" "abcabcbb" "bbbbb" "pwwkew" "" ]
 * 
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
  let current = ''
  let a = 0

  // i: substring start, j: substring end
  
  for (let i=0; i < s.length; i++) {
      let ch = s[i]
      let j = current.indexOf(ch)
      
      if (j > -1) {
          a = Math.max(a, current.length)
          current = current.substr(j+1) + ch
      }
      
      else current += ch
  }
  
  return Math.max([a, current.length])
};

File: LeetCode/3.LengthOfLongestSubstring.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        w = '' # our sliding window
        m = 0 # max
        
        '''
        i: substring start index
        j: index of a repeated character in our window
        '''
        for c in s:
            
            '''
            is the current character already in our sliding window?
            '''
            j = w.find(c)
            
            if j > -1:
                m = max(len(w), m)
                w = w[j+1:] + c
                
            else:
                w += c
                
        m = max(len(w), m)
        
        return m
        